src/audio/template_comp/template.c

```diff
@@ -0,0 +1,210 @@
+ * macro __cold informs that the code that is non-critical
+ * is loaded to slower but large DRAM.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+__cold static int template_comp_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct template_comp_comp_data *cd;
+
+	comp_info(dev, "template_comp_init()");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+	return 0;
+}
+
+/**
+ * template_comp_process() - The audio data processing function.
+ * @mod: Pointer to module data.
+ * @sources: Pointer to audio samples data sources array.
+ * @num_of_sources: Number of sources in the array.
+ * @sinks: Pointer to audio samples data sinks array.
+ * @num_of_sinks: Number of sinks in the array.
+ *
+ * This is the processing function that is called for scheduled
+ * pipelines. The processing is controlled by the enable switch.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+static int template_comp_process(struct processing_module *mod,
+				 struct sof_source **sources,
+				 int num_of_sources,
+				 struct sof_sink **sinks,
+				 int num_of_sinks)
+{
+	struct template_comp_comp_data *cd =  module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct sof_source *source = sources[0];
+	struct sof_sink *sink = sinks[0];
+	int frames = source_get_data_frames_available(source);
+	int sink_frames = sink_get_free_frames(sink);
+
+	comp_dbg(dev, "template_comp_process()");
+
+	frames = MIN(frames, sink_frames);
+	if (cd->enable)
+		/* Process the data with the channels swap example function. */
+		return cd->template_comp_func(mod, source, sink, frames);
+
+	/* Just copy from source to sink. */
+	source_to_sink_copy(source, sink, true, frames * cd->frame_bytes);
+	return 0;
+}
+
+/**
+ * template_comp_prepare() - Prepare the component for processing.
+ * @mod: Pointer to module data.
+ * @sources: Pointer to audio samples data sources array.
+ * @num_of_sources: Number of sources in the array.
+ * @sinks: Pointer to audio samples data sinks array.
+ * @num_of_sinks: Number of sinks in the array.
+ *
+ * Function prepare is called just before the pipeline is started. In
+ * this case the audio format parameters are for better code performance
+ * saved to component data to avoid to find out them in process. The
+ * processing function pointer is set to process the current audio format.
+ *
+ * Return: Value zero if success, otherwise error code.
+ */
+static int template_comp_prepare(struct processing_module *mod,
+				 struct sof_source **sources, int num_of_sources,
+				 struct sof_sink **sinks, int num_of_sinks)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	enum sof_ipc_frame source_format;
+	int i;
+
+	comp_dbg(dev, "template_comp_prepare()");
+
+	if (num_of_sources != 1 || num_of_sinks != 1)
```

Comment:
add a comment - there may be more than 1 input and/or output.

Reply:
@singalsu can you update incrementally, will merge this.

Reply:
Yep, I'll do another PR to address Marcin's additions.